### PR TITLE
Pipelined executor

### PR DESCRIPTION
**Pull request by sk1p**
This code is not fully benchmarked and battle tested yet, but works well together with corresponding changes in LiberTEM-live in the DECTRIS branch. Not optimized for offline processing at all, yet.

## TODO

- [x] Corresponding changes in LiberTEM-live: port merlin medipix acquisition object to new API, which might mean we need to update this PR again
- [x] Sometimes, in case of an exception in the UDF, the executor appears to hang - needs to be investigated
- [x] Comprehensive test coverage
- [x] Look into error behavior: possibly don't drain (only `close`) the response queue from the worker exception handler

## Contributor Checklist:

* [x] I have added or updated my entry in [the creators.json file](https://github.com/LiberTEM/LiberTEM/blob/master/packaging/creators.json)
* [x] I have added [a changelog entry](https://github.com/LiberTEM/LiberTEM/tree/master/docs/source/changelog) for my contribution
* [x] I have added/updated documentation for all user-facing changes
* [x] I have added/updated test cases
* [ ] I have included the [rebuilt production build of the client](https://libertem.github.io/LiberTEM/contributing.html?#building-the-client) (only if changes were made to the GUI)

## Reviewer Checklist:

* [x] `/azp run libertem.libertem-data` passed
* [ ] No import of GPL code from MIT code

### CODE REVIEW

**Comment by matbryan52** [comment]
Some feedback from testing:

- I was having a lot of test runs hang at the final call to `executor.close()`. This is using an executor which is a session fixture therefore shared between all tests. I think I've tracked this down to calling `qs.request.close()` before `p.join()` in `PipelinedExecutor.close()`. When I move the queue closure to the end of the loop it seems to finish up reliably every time.
- I've been working through how errors are handled and I believe that it's possible to leave the response queue in a bad state if a particular function / UDF run raises an error. It seems like you immediately `raise RuntimeError` on the main node as soon as we get an error message from a worker, but this can leave a bunch of tasks in flight which are going to fill up the response queue. If we then re-use the executor for another UDF run then the first responses we will recieve won't be from the same run. The only places where the queues are explicitly drained (that I can find) is when we get an unhandled error on a worker and we are cleaning up or we call `executor.close()`. Can reproduce with something like this:

```python
def test_udf_exception_queued(pipelined_ex):
    executor = pipelined_ex
    ctx = Context(executor=executor)

    data = np.random.randn(16, 16, 128, 128)
    ds = ctx.load("memory", data=data, num_partitions=16)

    error_udf = RaisesUDF()  # raises an error
    with pytest.raises(RuntimeError):  # raised by executor as expected
        ctx.run_udf(dataset=ds, udf=error_udf)

    # Fails immediately on run_udf because queue is in bad state
    normal_udf = SumUDF()
    ctx.run_udf(dataset=ds, udf=normal_udf)
    # Error is raised during the task dispatch loop when we check if any tasks completed yet
```

I think this could be fixed by having a try/finally block on `run_tasks` which explicitly drains the response (and request?) queues for all workers, but without closing them. For me this would make the behaviour consistent with other executors which can be re-used in case of errors (i.e. in a notebook).

**Comment by sk1p** [comment]
> * I was having a lot of test runs hang at the final call to `executor.close()`. This is using an executor which is a session fixture therefore shared between all tests. I think I've tracked this down to calling `qs.request.close()` before `p.join()` in `PipelinedExecutor.close()`. When I move the queue closure to the end of the loop it seems to finish up reliably every time.

Yeah - I think we can move the closing down, that should be fine.

> * I've been working through how errors are handled and I believe that it's possible to leave the response queue in a bad state if a particular function / UDF run raises an error. It seems like you immediately `raise RuntimeError` on the main node as soon as we get an error message from a worker, but this can leave a bunch of tasks in flight which are going to fill up the response queue. If we then re-use the executor for another UDF run then the first responses we will recieve won't be from the same run. The only places where the queues are explicitly drained (that I can find) is when we get an unhandled error on a worker and we are cleaning up or we call `executor.close()`. Can reproduce with something like this:
> 
> ```python
> def test_udf_exception_queued(pipelined_ex):
>     executor = pipelined_ex
>     ctx = Context(executor=executor)
> 
>     data = np.random.randn(16, 16, 128, 128)
>     ds = ctx.load("memory", data=data, num_partitions=16)
> 
>     error_udf = RaisesUDF()  # raises an error
>     with pytest.raises(RuntimeError):  # raised by executor as expected
>         ctx.run_udf(dataset=ds, udf=error_udf)
> 
>     # Fails immediately on run_udf because queue is in bad state
>     normal_udf = SumUDF()
>     ctx.run_udf(dataset=ds, udf=normal_udf)
>     # Error is raised during the task dispatch loop when we check if any tasks completed yet
> ```

Thanks for the reproducer!

> I think this could be fixed by having a try/finally block on `run_tasks` which explicitly drains the response (and request?) queues for all workers, but without closing them. For me this would make the behaviour consistent with other executors which can be re-used in case of errors (i.e. in a notebook).

Good call, yeah, keeping the executor in a good state after exceptions is the right behaviour, I agree. I think the `request` queues should only be drained on the worker itself, as the workers expect an `END_TASK` message at the end of processing a task using `RUN_TASK`. If they don't get this message, they will be left in a confused state and probably die. So the queue should always be drained by the process that usually reads from the given queue.

I think there is still error handling missing when a process of the pool dies unexpectedly, in which case it's possible that the queue(s) are in an invalid state (if the worker crashed). In the long run, we should detect this and just start a new pool of workers.

**Comment by matbryan52** [comment]
> @matbryan52 just for coordination, what is the status? Can you push your current version with the failing tests etc.? I could work on the queue draining / synchronization issues if needed, or let me know if you'd like to continue

You can push freely, no worries. As I said I was off for the end of last week and today have been working on something for `blobfinder`, so I haven't yet got around to making any changes to the executor itself. The comment I made above is more-or-less where I am.

**Comment by sk1p** [comment]
> You can push freely, no worries. As I said I was off for the end of last week and today have been working on something for `blobfinder`, so I haven't yet got around to making any changes to the executor itself. The comment I made above is more-or-less where I am

Ok, thanks!

**Review by uellue — CHANGES_REQUESTED** [review]
Looks great! 🚀 Just a few minor points, plus some items that @sk1p and I discussed in person.

File: tests/executor/test_pipelined.py [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+    ctx = Context(executor=executor)
+
+    data = np.random.randn(16, 16, 128, 128)
+    ds = ctx.load("memory", data=data, num_partitions=16)
+
+    error_udf = RaisesUDF()  # raises an error
+    with pytest.raises(RuntimeError):  # raised by executor as expected
+        ctx.run_udf(dataset=ds, udf=error_udf)
+
+    # Fails immediately on run_udf because queue is in bad state
```
[the remark this answers is not in the document]
Comment (sk1p) [reply tests/executor/test_pipelined.py:74]: Yup, will be removed.

File: tests/executor/test_pipelined.py [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+        )
+
+        def _break(a, b, c):
+            raise RuntimeError("stuff is broken, can't do it.")
+
+        def _do_patch_worker():
+            # XXX this completely destroys the workers ability to properly
+            # function, but that's okay because it's a throwaway process pool
+            # anyways:
+            pipelined.worker_run_function = _break
```
[the remark this answers is not in the document]
Comment (sk1p) [reply tests/executor/test_pipelined.py]: Done

File: tests/executor/test_pipelined.py [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+
+    error_udf = RaisesUDF()  # raises an error
+    with pytest.raises(RuntimeError):  # raised by executor as expected
+        ctx.run_udf(dataset=ds, udf=error_udf)
+
+    # Fails immediately on run_udf because queue is in bad state
+    normal_udf = SumUDF()
+    ctx.run_udf(dataset=ds, udf=normal_udf)
+    # Error is raised during the task dispatch loop when we check if any tasks
+    # completed yet
```
[the remark this answers is not in the document]
Comment (sk1p) [reply tests/executor/test_pipelined.py]: Done

**Review by uellue — APPROVED** [review]
💯

**Comment by sk1p** [comment]
I just noticed that the method to get the `TaskCommHandler` is still called `get_controller`, so we might need to do another round of updates

File: src/libertem/executor/pipelined.py [hunk 4]
```diff
@@ -0,0 +1,857 @@
...
+                                assert header_type not in (
+                                    "RUN_TASK", "SCATTER", "RUN_FUNCTION",
+                                    "CLEANUP", "SHUTDOWN", "WARMUP",
+                                )
+                                if header_type == "END_TASK":
+                                    break
+            elif header_type == "SCATTER":
+                # FIXME: array data could be transferred and stored in SHM instead
+                key = header["key"]
+                assert key not in work_mem
```
[the remark this answers is not in the document]
Comment (sk1p) [reply src/libertem/executor/pipelined.py:286]: Good point, I'll have a closer look at other `assert`s, too.

File: src/libertem/executor/pipelined.py [hunk 5]
```diff
@@ -0,0 +1,857 @@
...
+                        import libertem.preload  # NOQA
+                        with env.enter():
+                            pass
+            else:
+                queues.response.put({
+                    "type": "ERROR",
+                    "error": f"unknown message {header}",
+                    "worker_id": worker_idx,
+                })
+                continue
```
[the remark this answers is not in the document]
Comment (sk1p) [reply src/libertem/executor/pipelined.py:318]: Yeah - in that case, something is de-synchronized between worker and main process. I'll change it to `raise`.

File: src/libertem/executor/pipelined.py [hunk 6]
```diff
@@ -0,0 +1,857 @@
...
+                assert key not in work_mem
+                value = header["value"]
+                work_mem[key] = value
+                continue
+            elif header_type == "RUN_FUNCTION":
+                with attach_to_parent(header["span_context"]):
+                    worker_run_function(header, queues, worker_idx)
+            elif header_type == "CLEANUP":
+                key = header["key"]
+                assert key in work_mem
```
[the remark this answers is not in the document]
Comment (sk1p) [reply src/libertem/executor/pipelined.py:295]: :+1: on both points - changed.

File: src/libertem/executor/pipelined.py [hunk 7]
```diff
@@ -0,0 +1,857 @@
...
+        this limit (in detail: each worker's startup time is limited by this timeout).
+        In seconds.
+
+    cleanup_timeout
+        When cleaning up using :meth:`PipelinedExecutor.close`, give up after
+        this limit. In seconds.
+
+    early_setup
+        Callable that will be run very early on each worker process. Useful for
+        custom warmup code or testing.
```
[the remark this answers is not in the document]
Comment (sk1p) [reply src/libertem/executor/pipelined.py:567]: At least it is possible that there will be other "environment preparation" things that need to be executed before this function. Also, all modules imported from the executor are already imported when this function is run, so 'first' is also maybe a bit misleading. I'll try to reword this to clarify.

File: src/libertem/executor/pipelined.py [hunk 8]
```diff
@@ -0,0 +1,857 @@
...
+            self._validate_worker_state()
+            task_comm_handler.start()
+            span = trace.get_current_span()
+            span_context = span.get_span_context()
+
+            for qs, _ in self._pool.workers:
+                qs.request.put({
+                    "type": "WARMUP",
+                    "span_context": span_context,
+                })
```
[the remark this answers is not in the document]
Comment (sk1p) [reply src/libertem/executor/pipelined.py:668]: Good point - I don't think it has to run for every `run_udf` - I'll move this to `_start_pool` to also properly attribute the timings to the startup phase. Yeah, `import`ing things that are already in `sys.modules` is very cheap (<100ns), the `with env.enter()` can be a bit more expensive though.

I think I put it in `run_tasks` because I also wanted to warm up the whole I/O pipeline, in case we have `numba` compiled code in there. This of course only works if we know which `DataSet` / `Acquisition` instance will be used. I didn't get to implement this yet, so i'll defer this for now and move warmup to `_start_pool`.

File: src/libertem/executor/pipelined.py [hunk 9]
```diff
@@ -0,0 +1,857 @@
...
+
+            for task_idx, task in enumerate(tasks):
+                in_flight += 1
+                id_to_task[task_idx] = task
+
+                worker_idx = task_idx % self._pool.size
```
[the remark this answers is not in the document]
Comment (sk1p) [reply src/libertem/executor/pipelined.py:674]: I think both the threading and the mulitprocessing queue implement `qsize()` to get the approximate queue size - so it's not guaranteed to be correct, but for this purpose that should be enough. I did not check overhead.

For the first version I'd like to keep the simple "static" scheduling, but I'll add a note in the code to indicate that it can be improved. Maybe something for 0.11?

[the remark this answers is not in the document]
Comment (matbryan52) [reply src/libertem/executor/pipelined.py:674]: Agreed :+1:

File: src/libertem/executor/pipelined.py [hunk 10]
```diff
@@ -0,0 +1,857 @@
...
+                "type": "SCATTER",
+                "key": key,
+                "value": obj,
+            })
+        yield key
+        for qs, p in self._pool.workers:
+            qs.request.put({
+                "type": "CLEANUP",
+                "key": key,
+            })
```
[the remark this answers is not in the document]
Comment (sk1p) [reply src/libertem/executor/pipelined.py:849]: >The behaviour of .scatter() is quite different between executors:
[...]
> Might be worth either documenting the different behaviours or harmonising the APIs

Yeah, the idea is that the object that is yielded from `scatter` is an opaque handle, which is completely implementation-specific, and can be fed back into other executor functions which will then resolve the handle to the "real" object underneath. Currently it's only used for the `params_handle` for `run_tasks`. For harmonizing, one could think about emulating the dask behavior of having a kind of wrapper object around the handle, which is then resolved on the remove end if it appears as function parameters. That would probably mean slightly changing the `run_function` (and related) APIs.

> nothing happens when leaving context, object remains available within the
cluster accessible via the future(s)

IIRC dask frees up the worker memory when all futures that reference it are GC'd, so we don't need to do explicit cleanup.

> One particular scenario is closing the pool/queue while inside this .scatter() context? Obviously closed workers will not hold any data so no problem but this method will crash or hang when exiting!

Yeah, that means the request queue is already closed when the "DELETE" message is sent, causing a "queue is closed" exception. I'll add a test case for this specific behavior and just not send the "DELETE" if the pool is closed - that memory is GC'd anyway if the worker is gone.

File: src/libertem/executor/pipelined.py [hunk 11]
```diff
@@ -0,0 +1,857 @@
...
+                raise RuntimeError(f"failed to run function: {response['error']}")
+            if not response["type"] == "RUN_FUNCTION_RESULT":
+                raise RuntimeError(f"invalid response type: {response['type']}")
+            result: T = response["result"]
+            return result
+
+    def run_function(self, fn: Callable[..., T], *args, **kwargs) -> T:
+        # FIXME: this is not concurrency-safe currently! beware!
+        with tracer.start_as_current_span("PipelinedExecutor.run_tasks"):
+            return self._run_function(fn, 0, *args, **kwargs)
```
[the remark this answers is not in the document]
Comment (sk1p) [reply src/libertem/executor/pipelined.py:822]: If everything is working as intended, the queues are all zero here, as we can't have concurrent `run_function` etc. _yet_. Once we do, I think we could have a very simple `get_least_busy_worker` function that is used in both `run_function` and `run_tasks`.

File: src/libertem/executor/pipelined.py [hunk 12]
```diff
@@ -0,0 +1,857 @@
...
+        with qs.response.get() as (response, _):
+            if response["type"] == "ERROR":
+                raise RuntimeError(f"failed to run function: {response['error']}")
+            if not response["type"] == "RUN_FUNCTION_RESULT":
+                raise RuntimeError(f"invalid response type: {response['type']}")
+            result: T = response["result"]
+            return result
+
+    def run_function(self, fn: Callable[..., T], *args, **kwargs) -> T:
+        # FIXME: this is not concurrency-safe currently! beware!
```
[the remark this answers is not in the document]
Comment (sk1p) [reply src/libertem/executor/pipelined.py:820]: Good catch, fixing.

File: src/libertem/executor/pipelined.py [hunk 13]
```diff
@@ -0,0 +1,857 @@
...
+    Called in case the worker encounters an unhandled exception.
+    """
+    while True:
+        try:
+            with queues.request.get(block=False):
+                continue
+        except WorkerQueueEmpty:
+            break
+    # close and join:
+    queues.request.close()
```
[the remark this answers is not in the document]
Comment (sk1p) [reply src/libertem/executor/pipelined.py:349]: You are right. I think this might also be an issue: I think we must not drain the response queue here, on the worker, as that would mean responses can get lost before they reach the main process (i.e. written by the worker process and directly drained out again)

This might need a change in the `WorkerQueue` interface, I'll need to think about this, but it should be possible to do backwards-compatible.

File: src/libertem/executor/pipelined.py [hunk 14]
```diff
@@ -0,0 +1,857 @@
...
+    has_cupy: bool = False,  # currently ignored, for convenience of passing **detect()
+) -> List[WorkerSpec]:
+    """
+    Takes the output of `libertem.utils.devices.detect`
+    and makes a plan for starting workers on them.
+
+    Parameters
+    ----------
+
+    cpus
```
[the remark this answers is not in the document]
Comment (sk1p) [reply src/libertem/executor/pipelined.py:502]: Maybe something for a follow-up PR? I also like to have `make_spec` as a `classmethod`, maybe something we can port to the dask executor.

File: src/libertem/executor/pipelined.py [hunk 15]
```diff
@@ -0,0 +1,857 @@
...
+    def close_resp_queue(self):
+        while True:
+            try:
+                with self._response_q.get(block=False):
+                    continue
+            except WorkerQueueEmpty:
+                break
+        self._response_q.close()
+
+    def get_worker_queues(self, idx: int) -> WorkerQueues:
```
[the remark this answers is not in the document]
Comment (sk1p) [reply src/libertem/executor/pipelined.py]: Done.

File: src/libertem/executor/pipelined.py [hunk 16]
```diff
@@ -0,0 +1,857 @@
...
+        ])
+
+    def close(self):
+        with tracer.start_as_current_span("PipelinedExecutor.close") as span:
+            for idx, (qs, p) in enumerate(self._pool.workers):
+                span.add_event("sending SHUTDOWN", {"idx": idx})
+                qs.request.put({
+                    "type": "SHUTDOWN",
+                    "span_context": span.get_span_context(),
+                })
```
[the remark this answers is not in the document]
Comment (sk1p) [reply src/libertem/executor/pipelined.py]: Yeah, added a check :+1:

File: src/libertem/executor/pipelined.py [hunk 17]
```diff
@@ -0,0 +1,857 @@
...
+        resources = {"compute": 1, "CPU": 1}
+        return WorkerSet([
+            Worker(
+                name="cpu-%d" % idx,
+                host="localhost",
+                resources=resources,
+                nthreads=1,
+            )
+            for idx, worker in enumerate(self._pool.workers)
+        ])
```
[the remark this answers is not in the document]
Comment (sk1p) [reply src/libertem/executor/pipelined.py]: Especially, `run_each_worker` needs an update to use the right key (i.e. it cannot operate on `enumerate(set_of_workers)`, but instead needs to work on the guaranteed-ordered `self._pool.workers`.

File: src/libertem/executor/pipelined.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (sk1p) [reply src/libertem/executor/pipelined.py]: WARMUP is now moved to `_start_pool`

File: src/libertem/executor/pipelined.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (sk1p) [reply src/libertem/executor/pipelined.py]: I changed the wording to "as early as possible", which should then be a guideline for future changes, too.

File: src/libertem/executor/pipelined.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (sk1p) [reply src/libertem/executor/pipelined.py:506]: Created an issue for this: https://github.com/LiberTEM/LiberTEM/issues/1294

**Comment by sk1p** [comment]
Ugh, so in https://dev.azure.com/LiberTEM/LiberTEM/_build/results?buildId=1890&view=logs&j=d6eb154b-b7d5-5cda-1949-6de727f160fb&t=7c489ebc-6405-5d83-03ef-18776bc3088b we get a failure, because the `TVIPSDataSet` relies on exception propagation from `run_function`. Should be easy enough to implement™

File: src/libertem/executor/pipelined.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (matbryan52) [reply src/libertem/executor/pipelined.py]: > I think we must not drain the response queue here, on the worker, as that would mean responses can get lost before they reach the main process
Good spot, I didn't think about that!

> but it should be possible to do backwards-compatible.
Agreed.

**Comment by sk1p** [comment]
@matbryan52 thanks for the thorough review! I think most points should be addressed now.

@uellue could you review the worker resources part? It should mirror the dask executor impl. `has_cupy` is then either passed in by the user to `make_spec`, or automatically handled by `detect`.

> we get a failure, because the `TVIPSDataSet` relies on exception propagation from `run_function`. Should be easy enough to implement™

I had to change exception handling a bit - now, exceptions are re-raised on the main process by pickling. TVIPS tests now run through.

File: src/libertem/executor/pipelined.py [hunk 18]
```diff
@@ -0,0 +1,857 @@
...
+
+            for qs, _ in self._pool.workers:
+                qs.request.put({
+                    "type": "WARMUP",
+                    "span_context": span_context,
+                })
+
+            for task_idx, task in enumerate(tasks):
+                in_flight += 1
+                id_to_task[task_idx] = task
```
[the remark this answers is not in the document]
Comment (matbryan52) [reply src/libertem/executor/pipelined.py:692]: Good points all. I would imagine some kind of system where each 'job' (udf task or other) has a uuid associated, and results are pulled from the queue or an intermediate cache by requesting the particular UUID. In truth this is essentially re-implementing `concurrent.futures` or `Dask.futures` and we might want to look if we can re-use some of that API one day?

File: src/libertem/executor/pipelined.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (sk1p) [reply src/libertem/executor/pipelined.py]: `drain` parameter is now available for `WorkerQueue.close`

**Comment by matbryan52** [comment]
> Ugh, so in https://dev.azure.com/LiberTEM/LiberTEM/_build/results?buildId=1890&view=logs&j=d6eb154b-b7d5-5cda-1949-6de727f160fb&t=7c489ebc-6405-5d83-03ef-18776bc3088b we get a failure, because the `TVIPSDataSet` relies on exception propagation from `run_function`. Should be easy enough to implement™

Raising the same error on the main node is a great feature to have :+1: . I didn't know about `tblib`.

**Comment by matbryan52** [comment]
> @matbryan52 thanks for the thorough review! I think most points should be addressed now.

No worries, I don't think I can 'approve' as I submitted a comment review only, but for me it's all good now :+1:

File: src/libertem/executor/pipelined.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (sk1p) [reply src/libertem/executor/pipelined.py:686]: >  I would imagine some kind of system where each 'job' (udf task or other) has a uuid associated, and results are pulled from the queue or an intermediate cache by requesting the particular UUID. 

Yeah exactly.

> In truth this is essentially re-implementing `concurrent.futures` or `Dask.futures` and we might want to look if we can re-use some of that API one day?

Agreed, using API from `concurrent.futures` would be really nice. I think the tricky part here is the synchronization between starting a task and sending the data to the correct queue - right now, this is handled by sharing a queue for synchronization (starting a task etc.) and sending data.

It _could_ be possible to re-model this to a pull-model, meaning when a task is ready for execution, it specifically calls back to the main process and requests the data for the given task, which then blocks until that data is available. I think with that model, it would be easier to re-use any kind of process pool or `concurrent.futures.Executor` kind of abstraction - on the other hand, there would have to be additional buffering on the main process, keeping the data around until it can be sent to the correct queue.

_Then_, if this buffering happens in shared memory already, it _could_ work (at least) as well as it does now - as long as this is a single-node executor - in the distributed case, it would really be nice to speculatively send data to its destination as soon as we get it on the main node. Kind of like the difference between a cut-through switch and one that uses store-and-forward.

With `dask.futures` the main issue is we need an "out-of-band" communication channel, that matches the IPC that is used for client<>scheduler<>worker communication.

**Comment by sk1p** [comment]
> No worries, I don't think I can 'approve' as I submitted a comment review only, but for me it's all good now +1

Great! (btw. I think you can go to the "files changed" tab, select "review changes" on the top right, and just submit an empty comment with "approve changes" to accomplish this)

**Review by matbryan52 — APPROVED** [review]
Reviewed and approved

File: src/libertem/executor/pipelined.py [hunk 19]
```diff
@@ -0,0 +1,901 @@
...
+        resources_by_kind = {
+            "CPU": {"compute": 1, "CPU": 1, "ndarray": 1},
+            "CUDA": {"compute": 1, "CUDA": 1},
+        }
+
+        def _resources_for_spec(worker_spec: WorkerSpec):
+            resources = resources_by_kind[worker_spec["device_kind"]]
+            if worker_spec["has_cupy"]:
+                resources["ndarray"] = 1
+            return resources
```
Comment (uellue) [inline src/libertem/executor/pipelined.py:791]: This looks correct to me. Once the executor can map tasks to the eligible workers we can see if it passes the corresponding tests of the `DaskJobExecutor`.

**Review by uellue — APPROVED** [review]
LGTM, merging!
